server/src/main/java/org/elasticsearch/common/blobstore/BlobContainer.java

```diff
@@ -194,4 +194,40 @@ void writeMetadataBlob(String blobName, boolean failIfAlreadyExists, boolean ato
      * @throws  IOException if there were any failures in reading from the blob container.
      */
     Map<String, BlobMetadata> listBlobsByPrefix(String blobNamePrefix) throws IOException;
+
+    /**
+     * Atomically sets the value stored at the given key to {@code updated} if the {@code current value == expected}.
+     * Keys not yet used start at initial value 0. Returns the current value (before it was updated).
+     *
+     * @param key key of the value to update
+     * @param expected the expected value
+     * @param updated the new value
+     * @return the value read from the register (before it was updated)
```

Comment:
Some comments:

* Should we use "compare and swap" rather than "exchange", since it is more mainstream?
* I find the return value a bit trappy. It seems to mean the caller can check if the CAS succeeded `iff returned value == expected value`. However, if two or more callers concurrently called the CAS with the same expected value, but different new values, it'd be impossible to know which one succeeded, while all of them think they succeeded. Right?

Reply:
> I find the return value a bit trappy. It seems to mean the caller can check if the CAS succeeded iff returned value == expected value. However, if two or more callers concurrently called the CAS with the same expected value, but different new values, it'd be impossible to know which one succeeded, while all of them think they succeeded. Right?

The updates are expected to be atomic, in the scenario that you describe the caller that "loses" the exchange should get the updated value and therefore consider the CAS as failed.

Reply:
1. I'm following the naming used by things like `AtomicReference`, which I believe is derived from the `CMPXCHG` x86 instruction.

2. If the first write succeeds then the second attempt will see a different value returned and know that it failed.

Reply:
No, the caller checks the return (pre-update) value. For instance if the current value is `1` and two callers try and do a `cax(1,2)` and `cax(1,3)` then these are the only possible executions:

```
cax(1,2) --> 1, success
cax(1,3) --> 2, failed
```

or

```
cax(1,3) --> 1, success
cax(1,2) --> 3, failed
```

Reply:
Ah I see now, thanks for the clarification! I think the `return the value read from the register (before it was updated)` threw me off because I thought in both cases of success/failure, it'd return `1`. Understanding now that this means the value found in the register and not the one provided to the cas function.